### PR TITLE
Rework similarities to use lucene's PerFieldSimilarityWrapper on MapperService

### PR DESCRIPTION
**Pull request by romseygeek**
Similarity providers for text fields are currently held directly on MappedFieldType,
and accessing them requires both a MapperService and a SimilarityService.  This 
makes things unnecessarily complicated (for example, Mapper.ParserContext needs 
to hold references to both), and clutters the field type interface with
a field that only makes sense for text-based mappers.

This PR refactors things so that the MapperService is responsible for providing a
Similarity for an index, using lucene's PerFieldSimilarityWrapper to make field
configurations possible.  Similarities for text fields are collected by visiting all
FieldMappers after a merge, and combining them appropriately.

### CODE REVIEW

**Comment by romseygeek** [comment]
If this works then we should also be able to do the same thing for the index, search and search quote analyzers, further simplifying the MappedFieldType class.

File: server/src/main/java/org/elasticsearch/index/mapper/Mapper.java [hunk 1]
```diff
@@ -172,6 +161,10 @@ public final String simpleName() {
      *  Both {@code this} and {@code mergeWith} will be left unmodified. */
     public abstract Mapper merge(Mapper mergeWith);
 
+    public void collectPerFieldResources(PerFieldResourceCollector collector) {
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/mapper/Mapper.java:164]: If I'm understanding correctly, it's still possible to set a similarity on a non-text field, but we now ignore it. Although I agree that similarities aren't nearly as useful for non-text fields, this behavior seems trappy and it's technically a breaking change.

My vote would be to maintain the previous behavior where similarities on types like `keyword` were respected. If we want to disallow setting similarities on non-text types, we could open a dedicated issue to discuss it.

**Review by jtibshirani — COMMENTED** [review]
I left one high-level comment, but haven't done a detailed review yet.

**Comment by romseygeek** [comment]
Superseded by #58439

**Pull request closed without merging**